File: app/src/main/java/com/example/tfg/Retrofit/WebService.java
package com.example.tfg.Retrofit;
import com.example.tfg.Retrofit.DataClases.Mesa;
import com.example.tfg.Retrofit.DataClases.Producto;
import com.example.tfg.Retrofit.DataClases.Reserva;
import com.example.tfg.Retrofit.DataClases.User;

import java.util.List;

import retrofit2.Call;
import retrofit2.Response;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

public interface WebService {
    //MESAS
    @GET("mesas")
    Call<List<Mesa>> getMesas();
    @GET("mesas/{id}")
    Call<Mesa> getMesa(@Path("id") String idMesa);
    @POST("mesas")
    Call<Mesa> createMesa(@Body Mesa newMesa);
    @PUT("mesas/{id}")
    Call<Mesa> updateMesa(@Path("id") String idMesa, @Body Mesa mesa);
    @DELETE("mesas/{id}")
    Call<Void> deleteMesa(@Path("id") String idMesa);
    @GET("mesas/ocupadas")
    Call<List<Mesa>> getMesasOcupadas();
    @PUT("mesas/{id}/estado")
    Call<Mesa> updateEstadoMesa(@Path("id") String idMesa, @Body Mesa mesa);
    @POST("mesas/{id}/productos")
    Call<Mesa> addProductoAMesa(@Path("id") String idMesa, @Body Producto producto);
    @DELETE("mesas/{id}/productos")
    Call<Mesa> removeProductoDeMesa(@Path("id") String idMesa, @Body Producto producto);


    //RESERVAS
    @GET("reservas")
    Call<List<Reserva>> getReservas();

    @GET("reservas/{id}")
    Call<Reserva> getReserva(@Path("id") String idReserva);

    @POST("reservas")
    Call<Reserva> createReserva(@Body Reserva newReserva);

    @PUT("reservas/{id}")
    Call<Reserva> updateReserva(@Path("id") String idReserva, @Body Reserva reserva);

    @DELETE("reservas/{id}")
    Call<Void> deleteReserva(@Path("id") String idReserva);

    @GET("reservas/fecha/{fecha}")
    Call<List<Reserva>> getReservasPorFecha(@Path("fecha") String fecha);

    //USUARIOS
    @POST("users/signup")
    Call<User> signUp(@Body User newUser);

    @POST("users/signin")
    Call<User> signIn(@Body User user);
}

File: app/src/main/java/com/example/tfg/Retrofit/WebService.kt
package com.example.tfg.Retrofit

import com.example.tfg.Retrofit.DataClases.Critica
import com.example.tfg.Retrofit.DataClases.Reserva
import com.example.tfg.Retrofit.Response.CriticaResponse
import com.example.tfg.Retrofit.Response.ReservaResponse
import com.example.tfg.Retrofit.Response.UserResponse
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path

interface WebService {
    data class Request(
        val email: String,
        val password: String,
        val nombre: String,
        val apellidos: String,
        val telefono: String,
        val rol: String,
        val idUsuario: String
    )

    @GET("/reservas")
    suspend fun getReservas(): Response<ReservaResponse>

    @POST("/hacerReservas")
    suspend fun createReserva(
        @Body reserva: Reserva
    ): Response<ReservaResponse>

    @PUT("/modificarReserva/{idReserva}")
    suspend fun updateReserva(
        @Path("idReserva") idReserva: String,
        @Body reserva: Reserva
    ): Response<ReservaResponse>

    @DELETE("/cancelaReserva/{idReserva}")
    suspend fun deleteReserva(
        @Path("idReserva") idReserva: String
    ): Response<ReservaResponse>

    @GET("/reservas/{idReserva}")
    suspend fun getReserva(
        @Path("idReserva") idReserva: String
    ): Response<ReservaResponse>

    @GET("/reservas/fecha/{fecha}")
    suspend fun getReservaPorFecha(
        @Path("fecha") fecha: String
    ): Response<ReservaResponse>

    @POST("/signup")
    suspend fun signUp(@Body request: Request): Response<UserResponse>

    @POST("/signin")
    suspend fun signIn(@Body request: Request): Response<UserResponse>

    @GET("/criticas")
    suspend fun obtenerCriticas(): Response<CriticaResponse>

    @GET("/criticas/{idCritica}")
    suspend fun obtenerCritica(@Path("idCritica") idCritica: Int): Response<CriticaResponse>

    @POST("/hacerCriticas")
    suspend fun crearCritica(@Body critica: Critica): Response<CriticaResponse>

    @PUT("/actualizarCritica/{idCritica}")
    suspend fun actualizarCritica(
        @Path("idCritica") idCritica: String,
        @Body critica: Critica
    ): Response<CriticaResponse>

    @DELETE("/eliminarCritica/{idCritica}")
    suspend fun eliminarCritica(@Path("idCritica") idCritica: String): Response<CriticaResponse>

    @GET("/criticas/usuario/{idUsuario}")
    suspend fun obtenerCriticasUsuario(@Path("idUsuario") idUsuario: Int): Response<CriticaResponse>
}

